EnoEngine/FlagSubmissionEndpoint.cs

```diff
@@ -25,34 +25,55 @@ namespace EnoEngine.FlagSubmission
 {
     class FlagSubmissionEndpoint
     {
-        private static readonly Dictionary<long, Channel<(Flag Flag, TaskCompletionSource<FlagSubmissionResult> FeedbackSource)>> Channels =
+        private readonly Dictionary<long, Channel<(Flag Flag, TaskCompletionSource<FlagSubmissionResult> FeedbackSource)>> Channels =
             new Dictionary<long, Channel<(Flag, TaskCompletionSource<FlagSubmissionResult>)>>();
+        private readonly Dictionary<long, TeamFlagSubmissionStatistic> SubmissionStatistics = new Dictionary<long, TeamFlagSubmissionStatistic>();
         private const int MaximumLineLength = 100;
         private readonly TcpListener ProductionListener = new TcpListener(IPAddress.IPv6Any, 1337);
         private readonly TcpListener DebugListener = new TcpListener(IPAddress.IPv6Any, 1338);
         private readonly ILogger Logger;
         private readonly JsonConfiguration Configuration;
         private readonly IServiceProvider ServiceProvider;
-        private readonly EnoStatistics Statistics;
+        private readonly EnoStatistics EnoStatistics;
 
         public FlagSubmissionEndpoint(IServiceProvider serviceProvider, ILogger<FlagSubmissionEndpoint> logger, JsonConfiguration configuration, EnoStatistics statistics)
         {
             Logger = logger;
             Configuration = configuration;
-            Statistics = statistics;
+            EnoStatistics = statistics;
             ProductionListener.Server.SetSocketOption(SocketOptionLevel.IPv6, SocketOptionName.IPv6Only, false);
             DebugListener.Server.SetSocketOption(SocketOptionLevel.IPv6, SocketOptionName.IPv6Only, false);
             ServiceProvider = serviceProvider;
             foreach (var team in configuration.Teams)
             {
                 Channels[team.Id] = Channel.CreateBounded<(Flag, TaskCompletionSource<FlagSubmissionResult>)>(new BoundedChannelOptions(100) { SingleReader = false, SingleWriter = false });
+                SubmissionStatistics[team.Id] = new TeamFlagSubmissionStatistic(team.Id);
+            }
+        }
+
+        public async Task LogSubmissionStatistics(long teamId, CancellationToken token)
+        {
+            var statistic = SubmissionStatistics[teamId];
+            while (!token.IsCancellationRequested)
+            {
+                var okFlags = Interlocked.Exchange(ref statistic.OkFlags, 0);
+                var oldFlags = Interlocked.Exchange(ref statistic.OldFlags, 0);
+                var ownFlags = Interlocked.Exchange(ref statistic.OwnFlags, 0);
+                var duplicateFlags = Interlocked.Exchange(ref statistic.DuplicateFlags, 0);
+                var invalidFlags = Interlocked.Exchange(ref statistic.InvalidFlags, 0);
+                EnoStatistics.FlagSubmissionStatisticsMessage(teamId, okFlags, duplicateFlags, oldFlags, invalidFlags, ownFlags);
+                await Task.Delay(5000);

```

Comment:
Btw what is this delay for?

Reply:
To Only Submit the statistics every 5 seconds